### PR TITLE
[Internal] Diagnostics: Adds spec for CosmosDiagnostics compaction summary mode

### PR DESCRIPTION
**Pull request by NaluTripician**
## Summary

Adds a spec sheet for the diagnostics compaction feature that introduces a `DiagnosticsVerbosity` option (`Detailed` vs `Summary`) to reduce unbounded diagnostics output size during high-retry scenarios (429 throttling, transient failures, cross-region failovers).

**Spec location:** `specs/diagnostics-compaction.md`

## Key Design Decisions

- **Summary mode** groups requests by region, keeps first/last in full detail, deduplicates middle requests with aggregate stats (count, RU, min/max/P50/avg latency)
- **Default is `Detailed`** — no behavioral change for existing users
- **Configuration:** `CosmosClientOptions.DiagnosticsVerbosity` + per-request `RequestOptions.DiagnosticsVerbosity` override + env vars
- **Max summary size:** 8KB default (4KB min), with truncated fallback
- **Compaction is serialization-only** — in-memory `ITrace` tree is unchanged

## Reference

Modeled after the Rust SDK's approach: Azure/azure-sdk-for-rust#3592

## Work Items

| # | Work Item | Description |
|---|-----------|-------------|
| 1 | Enum & Options Plumbing | `DiagnosticsVerbosity` enum, properties on `CosmosClientOptions`/`RequestOptions`, env var support |
| 2 | Summary Computation Engine | Trace tree walk, region grouping, first/last/aggregated groups |
| 3 | Serialization Integration | Branch `ToString()` on verbosity, size enforcement, truncated fallback |
| 4 | Contract Updates | Update `ContractEnforcementTests` baselines |
| 5 | Unit Tests | 21 tests covering summary engine, verbosity flow, edge cases |
| 6 | Integration Tests | Emulator-based tests for real operations with summary mode |
| 7 | Baseline Tests | Golden-file JSONs for summary mode serialization stability |

**Please review the spec and leave feedback before implementation begins.**

#5325

### CODE REVIEW

File: specs/diagnostics-compaction.md [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+// File: Microsoft.Azure.Cosmos/src/RequestOptions/RequestOptions.cs
+public class RequestOptions
+{
+    // ... existing properties ...
+
+    /// <summary>
+    /// Gets or sets the diagnostics verbosity for this specific request.
+    /// When null, the client-wide default from <see cref="CosmosClientOptions.DiagnosticsVerbosity"/> is used.
+    /// </summary>
+    public DiagnosticsVerbosity? DiagnosticsVerbosity { get; set; }
```
Comment (FabianMeiswinkel) [inline specs/diagnostics-compaction.md]: Why is this on RequestOptions - when it only impacts serialization you would only need it when calling CosmosDiagsnotcis.ToString (or ToJson) and add the verbosity there?

File: specs/diagnostics-compaction.md [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+{
+    // ... existing members ...
+
+    /// <summary>
+    /// Gets or sets the verbosity level used when serializing these diagnostics via ToString().
+    /// This is set automatically from CosmosClientOptions but can be
+    /// overridden by the caller before calling ToString(), or by using the
+    /// ToString(DiagnosticsVerbosity) overload.
+    /// </summary>
+    public DiagnosticsVerbosity Verbosity { get; set; } = DiagnosticsVerbosity.Detailed;
```
Comment (FabianMeiswinkel) [inline specs/diagnostics-compaction.md]: Why even have this property when there is a ToString overload accepting it? IMO this is not needed and introduces weird race conditions - just drop it - ToString() defaults to Detailed (backwards compatibility) and anyone else can use the new overloads.

**Review by FabianMeiswinkel — DISMISSED** [review]
LGTM except for the DiagnosticsVerbosity on CosmsoDiagnostics

File: specs/diagnostics-compaction.md [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    /// <param name="verbosity">The verbosity level to use for serialization.</param>
+    /// <returns>A JSON string with diagnostics at the requested verbosity level.</returns>
+    public abstract string ToString(DiagnosticsVerbosity verbosity);
+
+    /// <summary>
+    /// Returns a JSON representation of diagnostics using the specified verbosity.
+    /// </summary>
+    /// <param name="verbosity">The verbosity level to use for serialization.</param>
+    /// <returns>A JSON string with diagnostics at the requested verbosity level.</returns>
+    public abstract string ToJsonString(DiagnosticsVerbosity verbosity);
```
Comment (kirankumarkolli) [inline specs/diagnostics-compaction.md]: Is it needed?

File: specs/diagnostics-compaction.md [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+    /// </summary>
+    public DiagnosticsVerbosity DiagnosticsVerbosity { get; set; } = DiagnosticsVerbosity.Detailed;
+
+    /// <summary>
+    /// Gets or sets the maximum size in bytes for Summary mode diagnostic output.
+    /// If the summary output exceeds this limit, a truncated indicator is returned.
+    /// Default: 8192 (8 KB). Minimum: 4096 (4 KB).
+    /// Can also be set via the AZURE_COSMOS_DIAGNOSTICS_MAX_SUMMARY_SIZE environment variable.
+    /// </summary>
+    public int MaxDiagnosticsSummarySizeBytes { get; set; } = 8192;
```
Comment (kirankumarkolli) [inline specs/diagnostics-compaction.md]: What's the impact of it on the fidelity of the troubleshooting context?

File: specs/diagnostics-compaction.md [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+## 1. Problem Statement
+
+`CosmosDiagnostics.ToString()` produces a JSON trace that grows **unboundedly** with retries. Each retry attempt creates a new child `ITrace` node containing a full `ClientSideRequestStatisticsTraceDatum` with complete `StoreResponseStatistics` and `HttpResponseStatistics` entries. In pathological scenarios (sustained 429 throttling, transient failures, cross-region failovers), a single operation's diagnostics can grow to hundreds of KB.
+
+**Impact:**
+- **Log truncation** — monitoring systems (Application Insights, Azure Monitor, etc.) silently drop oversized log entries
+- **Memory pressure** — large diagnostic strings increase GC overhead, especially at high throughput
+- **Readability** — operators cannot quickly extract signal from noise when hundreds of identical retry entries are listed
+
+**Example scenario:** A point read that encounters 50 retries due to 429 throttling in West US 2, then fails over to East US 2 with 10 more retries, produces ~60 full `StoreResponseStatistics` entries in the trace tree. With summary mode, this compacts to: first request + last request + 1 aggregated group per region.
```
Comment (kirankumarkolli) [inline specs/diagnostics-compaction.md]: Cross partition queries are other bigger source of issues.

File: specs/diagnostics-compaction.md [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+        Detailed = 0,
+
+        /// <summary>
+        /// Compacted diagnostic output optimized for log size constraints.
+        /// Groups requests by region. Keeps first and last request in full detail.
+        /// Deduplicates middle requests by (StatusCode, SubStatusCode) with
+        /// aggregate statistics (count, total RU, min/max/P50 latency).
+        /// Respects MaxDiagnosticsSummarySizeBytes limit.
+        /// </summary>
+        Summary = 1,
```
Comment (kirankumarkolli) [inline specs/diagnostics-compaction.md]: What's the guidance for customers on when to use what?

File: specs/diagnostics-compaction.md [hunk 7]
```diff
@@ -0,0 +1,407 @@
+# CosmosDiagnostics Compaction — Summary Mode
+
+## 1. Problem Statement
+
+`CosmosDiagnostics.ToString()` produces a JSON trace that grows **unboundedly** with retries. Each retry attempt creates a new child `ITrace` node containing a full `ClientSideRequestStatisticsTraceDatum` with complete `StoreResponseStatistics` and `HttpResponseStatistics` entries. In pathological scenarios (sustained 429 throttling, transient failures, cross-region failovers), a single operation's diagnostics can grow to hundreds of KB.
+
+**Impact:**
+- **Log truncation** — monitoring systems (Application Insights, Azure Monitor, etc.) silently drop oversized log entries
```
Comment (kirankumarkolli) [inline specs/diagnostics-compaction.md]: JSON is verbose, thouhgts on encoding which might help/

Comment (NaluTripician) [reply specs/diagnostics-compaction.md]: Thoughts from offline discussion: 
- Consider a hybrid format with JSON text readable summary + encoded information that contains more details for debugging. Have an easy way to decode the information for IcMs + general debugging. Decoding would have to also be available for customers. Possibly add a tool in the repo to do this. Could be standardized across SDKs.

Comment (NaluTripician) [reply specs/diagnostics-compaction.md]: Also important, how can we preserve important information such as which replicas are contacted/failing?

File: openspec/specs/diagnostics-and-observability/spec.md [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+
+## Interactions
+
+- **Handler Pipeline**: `DiagnosticsHandler` (#3) captures CPU/system usage. `TelemetryHandler` (#4) collects operation telemetry. See `handler-pipeline` spec.
+- **Retry Policies**: Each retry attempt adds a child trace node, making retries visible in diagnostics. See `retry-and-failover` spec.
+
+## References
+
+- Source: `Microsoft.Azure.Cosmos/src/Diagnostics/CosmosDiagnostics.cs`
+- Source: `Microsoft.Azure.Cosmos/src/Diagnostics/DiagnosticsVerbosity.cs`
```
Comment (NaluTripician) [inline openspec/specs/diagnostics-and-observability/spec.md:392]: **🟡 Recommendation** · Maintainability: Spec Accuracy

**Spec references files that don't exist yet**

The References section lists `DiagnosticsVerbosity.cs` and `DiagnosticsSummaryWriter.cs` as source files, but these are files **to be created** (per `design.md`). I verified via the GitHub API that neither file exists in the repository today.

Since this spec catalog is positioned as 'authoritative context for implementation' and 'living documentation,' AI agents reading this spec will attempt to reference or call APIs from these non-existent files. More broadly, the entire DiagnosticsVerbosity section (~200 lines of the 399-line spec) describes proposed behavior mixed in with existing behavior, with no visual distinction.

**Suggestion:** Either (a) add a clear `> :construction: PROPOSED — Not yet implemented` callout to all DiagnosticsVerbosity sections, or (b) defer adding them to the main spec until the implementation PR merges — the `openspec/changes/diagnostics-compaction/` artifacts already capture the full design.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: openspec/specs/retry-and-failover/spec.md [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+### Requirement: ResourceThrottleRetryPolicy
+
+The `ResourceThrottleRetryPolicy` SHALL handle HTTP 429 (rate-limited) responses by retrying with server-specified delays up to configurable limits.
+
+**Throttle retry parameter reference table:**
+
+| Parameter | Default |
+|-----------|---------|
+| Max retry attempts | 9 |
+| Max cumulative wait time | 60 seconds |
```
Comment (NaluTripician) [inline openspec/specs/retry-and-failover/spec.md:132]: **🟡 Recommendation** · Correctness: Inconsistent Default

**MaxRetryWaitTime default disagrees with client-and-configuration spec**

This table says 'Max cumulative wait time' defaults to **60 seconds**. However, `openspec/specs/client-and-configuration/spec.md` says `MaxRetryWaitTimeOnRateLimitedRequests` defaults to **30 seconds**.

I verified both values in the source:
- `RetryOptions.DefaultMaxRetryWaitTimeInSeconds = 30` (the public-facing default)
- `ResourceThrottleRetryPolicy.DefaultMaxWaitTimeInSeconds = 60` (internal policy fallback)

The effective customer-facing default is **30 seconds** since `CosmosClientOptions` always passes `RetryOptions` values to the policy constructor. The 60s value here cites the internal policy fallback, not what customers experience.

**Suggestion:** Update this table to say 30 seconds (matching the effective default), or add a note: '60s is the internal policy fallback; effective default is 30s per `CosmosClientOptions`.'

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: openspec/specs/diagnostics-and-observability/spec.md [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+
+**Scenario: Summary mode size under limit**
+When the serialized summary JSON fits within `MaxDiagnosticsSummarySizeBytes`,
+the system shall return the full summary JSON as-is.
+
+**Scenario: MaxDiagnosticsSummarySizeBytes minimum validation**
+When `CosmosClientOptions.MaxDiagnosticsSummarySizeBytes` is set below 4096,
+the system shall reject the value (minimum: 4096 bytes).
+
+**Scenario: Verbosity precedence**
```
Comment (NaluTripician) [inline openspec/specs/diagnostics-and-observability/spec.md:209]: **🟡 Recommendation** · Correctness: Contradictory Requirements

**Verbosity precedence contradicts 'Parameterless ToString always returns Detailed'**

This section defines a 3-tier precedence: (1) explicit `ToString(DiagnosticsVerbosity)` parameter → (2) `CosmosClientOptions.DiagnosticsVerbosity` → (3) default Detailed.

But line 169 says: 'Parameterless `ToString()` always returns Detailed, **regardless of** `CosmosClientOptions.DiagnosticsVerbosity` setting.'

These two requirements contradict each other. If parameterless `ToString()` always returns Detailed regardless of the client option, then `CosmosClientOptions.DiagnosticsVerbosity` has no effect — there's no call path that consults it (since `ToString(verbosity)` already has an explicit parameter, and `ToString()` ignores it).

**Suggestion:** Clarify the intent: either (a) parameterless `ToString()` respects `CosmosClientOptions.DiagnosticsVerbosity` as its default (remove the 'regardless' clause on line 169), or (b) `CosmosClientOptions.DiagnosticsVerbosity` only provides a convenience default for callers who construct the verbosity parameter from config (document this explicitly). As written, the option appears useless.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: openspec/changes/diagnostics-compaction/design.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+| `SummaryDiagnostics.cs` | Extend `CollectSummaryFromTraceTree()` to support region-grouped collection with ordering |
+| `ClientSideRequestStatisticsTraceDatum.cs` | Ensure `StoreResponseStatistics` and `HttpResponseStatistics` lists are accessible for summary computation |
+
+## Contract/Baseline Updates
+
+| File | Change |
+|------|--------|
+| `ContractEnforcementTests.cs` baseline | Update public API contract for new enum and properties |
+
+## Alternatives Considered
```
Comment (NaluTripician) [inline openspec/changes/diagnostics-compaction/design.md:76]: **🟡 Recommendation** · Completeness: Missing Alternative

**Hybrid encoding alternative from review discussion not captured**

@kirankumarkolli raised 'JSON is verbose, thoughts on encoding which might help?' and @NaluTripician replied with thoughts from offline discussion: 'Consider a hybrid format with JSON text readable summary + encoded information that contains more details for debugging.'

This is a substantive alternative approach that emerged from review, but it's not listed in the Alternatives Considered section. Future readers won't know this option was discussed and evaluated.

**Suggestion:** Add 'Alternative 4: Hybrid format (readable summary + encoded details)' with pros/cons and decision rationale, even if the decision is to defer. This preserves the design context from the review discussion.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: openspec/changes/diagnostics-compaction/proposal.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+## Proposed Approach
+
+Introduce a **`DiagnosticsVerbosity`** concept (modeled after [Azure/azure-sdk-for-rust#3592](https://github.com/Azure/azure-sdk-for-rust/pull/3592)) that controls how `CosmosDiagnostics.ToString()` serializes trace data:
+
+| Mode | Behavior | Use Case |
+|------|----------|----------|
+| **Detailed** (default) | Current behavior — full trace tree output | Debugging, development |
+| **Summary** | Region-grouped compaction with first/last + aggregated middle | Production logging, size-constrained environments |
+
+**Key design principle:** The in-memory representation (`ITrace` tree, `ClientSideRequestStatisticsTraceDatum`) stays **unchanged**. Compaction only happens at **serialization time** in the `TraceJsonWriter` path. This preserves full programmatic access to diagnostics data while reducing serialized output size.
```
Comment (NaluTripician) [inline openspec/changes/diagnostics-compaction/proposal.md:23]: **🟡 Recommendation** · Completeness: Scope Gap

**Cross-partition query diagnostics growth not addressed**

@kirankumarkolli noted 'Cross partition queries are other bigger source of issues.' The proposal focuses on retry-heavy scenarios (429 throttling, failover), but cross-partition queries can produce large diagnostics due to fan-out across many physical partitions — regardless of retries.

A query fanning out to 50 partitions produces 50 separate `ClientSideRequestStatisticsTraceDatum` entries even without any retries. Summary mode's region-grouping helps somewhat, but the compaction algorithm (first/last + aggregated middle) is designed for the retry pattern, not the fan-out pattern.

**Suggestion:** Add a 'Scope Limitations' note or expand 'Non-Goals' to acknowledge that cross-partition query fan-out diagnostics are not addressed in this phase, and note whether a future phase could address it.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

File: openspec/changes/diagnostics-compaction/proposal.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+1. Ship with `Detailed` as default in initial release
+2. Document `Summary` mode in SDK documentation and changelog
+3. Consider making `Summary` the default in a future major version after customer feedback
+
+## Non-Goals
+
+- Changing the in-memory `ITrace` tree structure
+- Modifying the `Detailed` mode output format
+- Adding new programmatic APIs beyond `ToString(DiagnosticsVerbosity)` overload
+- Per-request verbosity override via `RequestOptions` (can be added later)
```
Comment (NaluTripician) [inline openspec/changes/diagnostics-compaction/proposal.md:51]: **🟢 Suggestion** · Completeness: Replica Information

**Consider preserving replica/endpoint information in summary mode**

@NaluTripician noted in PR comments: 'Also important, how can we preserve important information such as which replicas are contacted/failing?' This concern isn't addressed in the current spec.

Summary mode compacts middle requests into aggregated stats, but replica-level detail (which specific replicas were contacted, which failed) is potentially lost. This information is critical for IcM debugging — knowing that replica X in region Y is the one failing is often the key diagnostic signal.

**Suggestion:** Add a requirement or open question about preserving replica/endpoint information in summary mode output. At minimum, the `First` and `Last` entries preserve this, but consider whether aggregated groups should include a distinct endpoint count or list.

---
<sub>⚠️ AI-generated review — may be incorrect. Agree? → resolve the conversation. Disagree? → reply with your reasoning.</sub>

**Comment by NaluTripician** [comment]
Closing until rust work is done
